src/pokemon.c

```diff
@@ -0,0 +1,5567 @@
+    u32 * expTable;
+
+    GF_ASSERT(monExpRate < 8);
+    GF_ASSERT(monLevel <= 101);
+
+    expTable = Heap_AllocFromHeap(0, 101 * 4);
+    LoadMonExperienceTable(monExpRate, expTable);
+
+    result = expTable[monLevel];
+    Heap_FreeToHeap(expTable);
+
+    return result;
+}
+
+u32 GetMonLevel (Pokemon * mon)
+{
+    return GetBoxMonLevel(&mon->box);
+}
+
+u32 GetBoxMonLevel (BoxPokemon * boxMon)
+{
+    int monSpecies;
+    u32 monExp;
+    BOOL reencrypt;
+
+    reencrypt = DecryptBoxMon(boxMon);
+    monSpecies = GetBoxMonData(boxMon, MON_DATA_SPECIES, 0);
+    monExp = GetBoxMonData(boxMon, MON_DATA_EXP, 0);
+
+    EncryptBoxMon(boxMon, reencrypt);
+
+    return GetMonSpeciesLevel(monSpecies, monExp);
+}
+
+u32 GetMonSpeciesLevel (u16 monSpecies, u32 monExp)
+{
+    int unused_0;
+    u32 monLevel;
+    u32 * unused_1;
+    PokemonPersonalData * monPersonalData = GetMonPersonalData(monSpecies, 0);
+
+    monLevel = GetMonPersonalDataLevel(monPersonalData, monSpecies, monExp);
+    FreeMonPersonalData(monPersonalData);
+
+    return monLevel;
+}
+
+u32 GetMonPersonalDataLevel (PokemonPersonalData * monPersonalData, u16 unused_monSpecies, u32 monExp)
+{
+    static u32 monExpTable[101];
+    int monExpRate, i;
+
+    monExpRate = GetMonPersonalDataAttribute(monPersonalData, MON_DATA_PERSONAL_EXP_RATE);
+    LoadMonExperienceTable(monExpRate, monExpTable);
+
+    for (i = 1; i < 101; i++) {
+        if (monExpTable[i] > monExp) {
+            break;
+        }
+    }
+
+    return i - 1;
+}
+
+u8 GetMonNature (Pokemon * mon)
+{
+    return GetBoxMonNature(&mon->box);
+}
+
+u8 GetBoxMonNature (BoxPokemon * boxMon)
+{
+    BOOL reencrypt;
+    u32 monPersonality;
+
+    reencrypt = DecryptBoxMon(boxMon);
+    monPersonality = GetBoxMonData(boxMon, MON_DATA_PERSONALITY, 0);
+
+    EncryptBoxMon(boxMon, reencrypt);
+
+    return GetNatureFromPersonality(monPersonality);
+}
+
+u8 GetNatureFromPersonality (u32 monPersonality)
+{
+    return (u8)(monPersonality % 25);
+}
+
+static const s8 sNatureStatAffinities[][5] = {
+    {0x0, 0x0, 0x0, 0x0, 0x0},
+    {0x1, -1, 0x0, 0x0, 0x0},
+    {0x1, 0x0, -1, 0x0, 0x0},
+    {0x1, 0x0, 0x0, -1, 0x0},
+    {0x1, 0x0, 0x0, 0x0, -1},
+    {-1, 0x1, 0x0, 0x0, 0x0},
+    {0x0, 0x0, 0x0, 0x0, 0x0},
+    {0x0, 0x1, -1, 0x0, 0x0},
+    {0x0, 0x1, 0x0, -1, 0x0},
+    {0x0, 0x1, 0x0, 0x0, -1},
+    {-1, 0x0, 0x1, 0x0, 0x0},
+    {0x0, -1, 0x1, 0x0, 0x0},
+    {0x0, 0x0, 0x0, 0x0, 0x0},
+    {0x0, 0x0, 0x1, -1, 0x0},
+    {0x0, 0x0, 0x1, 0x0, -1},
+    {-1, 0x0, 0x0, 0x1, 0x0},
+    {0x0, -1, 0x0, 0x1, 0x0},
+    {0x0, 0x0, -1, 0x1, 0x0},
+    {0x0, 0x0, 0x0, 0x0, 0x0},
+    {0x0, 0x0, 0x0, 0x1, -1},
+    {-1, 0x0, 0x0, 0x0, 0x1},
+    {0x0, -1, 0x0, 0x0, 0x1},
+    {0x0, 0x0, -1, 0x0, 0x1},
+    {0x0, 0x0, 0x0, -1, 0x1},
+    {0x0, 0x0, 0x0, 0x0, 0x0}
+};
+
+u16 GetNatureAdjustedStatValue (u8 monNature, u16 monStatValue, u8 statType)
+{
+    u16 result;
+
+    if ((statType < 0x1) || (statType > 0x5)) {
```

Comment:
Looks like a sentinel on making sure `statType` is one of Attack, Defense, Speed, SpAttack, or SpDefense. Could use some consts.